buffer/src/main/java/io/netty/buffer/AbstractByteBuf.java

```diff
@@ -1210,7 +1210,11 @@ public String toString(Charset charset) {
 
     @Override
     public String toString(int index, int length, Charset charset) {
-        return ByteBufUtil.decodeString(this, index, length, charset);
+        if (charset.equals(CharsetUtil.UTF_8)) {
+            return ByteBufUtil.readUtf8(this, index, length); // Fast utf-8 decoder
+        } else {
```

Comment:
I was confused at first but than I realized that you wanted me to keep the if check in `ByteBufUtil.decodeString`